vault-core/src/test/java/org/apache/jackrabbit/vault/packaging/integration/MandatoryNodeIT.java

```diff
@@ -0,0 +1,62 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.jackrabbit.vault.packaging.integration;
+
+import java.io.IOException;
+
+import javax.jcr.RepositoryException;
+import org.apache.jackrabbit.vault.packaging.PackageException;
+import org.junit.Test;
+
+
+public class MandatoryNodeIT extends IntegrationTestBase {
+ 
+    @Test
+    public void testMandatoryChildNodesReplace()
+            throws RepositoryException, IOException, PackageException {
+
+        // package set ups folder structure with mandatory child nodes
+        extractVaultPackage("/test-packages/wcm-rollout-config-1.zip");
+
+        assertNodeExists("/libs/msm/wcm/rolloutconfigs/activate/targetActivate");
+
+        assertNodeExists("/libs/msm/wcm/rolloutconfigs/deactivate/targetDeactivate");
+
+        assertNodeExists("/libs/msm/wcm/rolloutconfigs/default/contentUpdate");
+        assertNodeExists("/libs/msm/wcm/rolloutconfigs/default/contentCopy");
+        assertNodeExists("/libs/msm/wcm/rolloutconfigs/default/contentDelete");
+
+        // packages removes/replaces some mandatory child nodes
+        extractVaultPackage("/test-packages/wcm-rollout-config-2.zip");
+
+        // check if older nodes are replaced
+        assertNodeMissing("/libs/msm/wcm/rolloutconfigs/default/contentUpdate");
+        assertNodeMissing("/libs/msm/wcm/rolloutconfigs/default/contentDelete");
+        // check if newly added node exist
+        assertNodeExists("/libs/msm/wcm/rolloutconfigs/default/activate");
+        assertNodeExists("/libs/msm/wcm/rolloutconfigs/default/contentCopy");
+
+        // check if the only node is replaced
+        assertNodeMissing("/libs/msm/wcm/rolloutconfigs/activate/targetActivate");
+        assertNodeExists("/libs/msm/wcm/rolloutconfigs/activate/targetNewActivate");
+
+        // check if the only node is still there (although not part of 2nd package)
+        assertNodeExists("/libs/msm/wcm/rolloutconfigs/deactivate/targetDeactivate");
```

Comment:
If I modify the code in DocViewSAXImporter.endNode() to remove mandatory child nodes this one suprisingly does not throw a ViolationConstraintException in Oak. @reschke any idea why deleting a mandatory child node is accepted?

Reply:
https://issues.apache.org/jira/browse/OAK-9621

Reply:
maybe we can relax the check and only reject the removal of the mandatory childnode, if it's a single node (eg, for a mandatory nt:resource). I think the _at least 1_ behaviour is rare...

Reply:
I think JCR 2.0 spec is clear in this regard:
Mandatory residual child nodes definitions are invalid (and therefore not enforced),  mandatory non-residual child node definitions are valid. 

So FileVault should IMHO remove mandatory child nodes (even the last one) in case the definition is residual!